### PR TITLE
Refractor CtFileSystem

### PR DESCRIPTION
**Pull request by ForeverRainbow**
This is a refractor of CtFileSystem and changing everything to use `CtFileSystem::path`, afaik there are no more `Glib::ustring`'s being used for paths (hazzah!). 

I have added unit tests in `tests_filesystem.cpp` and currently everything is passing on linux at least

(Also yes I misclicked when writing this, oops)

### CODE REVIEW

**Comment by ForeverRainbow** [comment]
Tests failing on windows because the paths are being replaced properly (ironically), fixing that now

**Comment by giuspen** [comment]
Very nice work indeed

**Comment by txe** [comment]
Agree, though there are few issues. I write them down when finish reviewing.

**Comment by giuspen** [comment]
Thanks @txe!

File: future/src/ct/ct_filesystem.cc [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+        // https://stackoverflow.com/questions/42442189/how-to-open-spawn-a-file-with-glib-gtkmm-in-windows
+#ifdef _WIN32
+        ShellExecute(NULL, "open", folderpath.c_str(), NULL, NULL, SW_SHOWDEFAULT);
+#elif defined(__APPLE__)
+        std::vector<std::string> argv = { "open", folderpath.string() };
+Glib::spawn_async("", argv, Glib::SpawnFlags::SPAWN_SEARCH_PATH);
+#else
+        fs::path path("file://");
+        path += folderpath;
+        g_app_info_launch_default_for_uri(folderpath.c_str(), nullptr, nullptr);
```
Comment (txe) [inline future/src/ct/ct_filesystem.cc:197]: folderpath -> path

File: future/src/ct/ct_filesystem.cc [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+            count += remove_all(file);
+        }
+        remove(file);
+    }
+    remove(dir);
+    ++count;
+    return count;
+}
+
+std::string get_cherrytree_datadir()
```
Comment (txe) [inline future/src/ct/ct_filesystem.cc:236]: why not fs::path?

Comment (txe) [reply future/src/ct/ct_filesystem.cc:236]: I see `fs::path get_cherrytree_localedir()` with fs::path, but others are std::string

File: future/src/ct/ct_misc_utils.cc [hunk 3]
```diff
@@ -760,6 +728,17 @@ char* CtRgbUtil::set_rgb24str_from_str_any(const char* rgbStrAny, char* rgb24Str
     return rgb24StrOut;
 }
 
+std::vector<std::string> CtMiscUtil::split(const std::string& str, std::string_view sep) {
```
Comment (txe) [inline future/src/ct/ct_misc_utils.cc:731]: there is `str::split`, why one more?

File: future/src/ct/ct_misc_utils.h [hunk 4]
```diff
@@ -47,14 +47,22 @@ namespace CtCSV {
 }
 
 namespace CtMiscUtil {
+template<class ITER_T>
```
Comment (txe) [inline future/src/ct/ct_misc_utils.h:50]: there is `str::join`

File: future/src/ct/ct_filesystem.cc (hunk 2 above)
Comment (ForeverRainbow) [reply future/src/ct/ct_filesystem.cc:236]: Must have missed those, I'll correct

File: future/src/ct/ct_misc_utils.cc (hunk 3 above)
Comment (ForeverRainbow) [reply future/src/ct/ct_misc_utils.cc:731]: Because I didn't know about `str::split` is why xD

**Comment by txe** [comment]
Another big thing is using `_get_platform_path`, there are a few reasons why it's bad:
- it looks like a sign of bad code when it is applied every time to every path. There were only a few places where `get_proper_platform_filepath` was used, though I don't see which problem it was fixing.
- it actually breaks paths, from this PR
```
fs::path path("file://");
path += folderpath;
```
on win `//` will be replaced by `\\`. Also paths can contain network paths, they will be also broken.

I suggest to remove `_get_platform_path` and revert `get_proper_platform_filepath` (still don't know what they are for)


the code for `path path::extension() const` doesn't look good, can you rewrite it by using the same approach as 'stem'?  Also about `stem`, it returns `path` though `string` have more sense.

**Comment by ForeverRainbow** [comment]
`stem()` returns `path` because it could very easily be a `path`, consider `/foo/bar.txt`, `stem() -> /foo/bar` which is a path. Also the specification states that `stem()` returns a `path` object. I have avoided network paths using `path` precisely for that reason (I think I have used it in a few places which is a problem).

The problem with a separate function is it breaks the guarantee that `string()` returns a correctly formatted path, which is specified by the standard. Our version of `path` doesn't have to be completely compatible with the standard of course, it just might cause confusion if later we switch to a compatible one. Originally `get_proper_platform_filepath` was called on every call to `string()` but I moved it. Instead of making it separate I could move it back or I could specialise `CtSplittable` to build a path using native separators and use that?

**Comment by ForeverRainbow** [comment]
I think calling `_get_platform_path` would probably be better since otherwise theres just the issue again of formatting when not needed. The standard does define a `format` enum which can be `native`, `generic` or `auto` (implementation define), `generic` just uses `/` seperators whereas `native` is what we have now.

**Comment by ForeverRainbow** [comment]
Yeah I will revert to `get_proper_platform_filepath`, the only places I think its actually needed are with the `system` calls since powershell can handle posix paths anyway (but `std::system` used `cmd.exe` even on windows 10 because reasons) but I will do it tomorrow

**Comment by giuspen** [comment]
What is your timezone @ForeverRainbow?

**Comment by txe** [comment]
well, `/foo/bar.txt, stem() -> /foo/bar` is not correct,  `stem()` should return `bar`, you can check `std::fs::path` specification

**Comment by ForeverRainbow** [comment]
@txe Yeah your right it should (will check that this one does), so it would make a bit more sense to return `string()` but since the standard specifies `path` it would make life easier for swapping it out later. 

@giuspen GMT

**Comment by ForeverRainbow** [comment]
OK I have moved the `_get_platform_path` call to `native()`, really `string()` should return `native()` (according to the standard) but since the change is compatible with any standard API it doesn't matter afaik (we just have to remember that `string()` does not return a windows correct path on windows). I think I have changed it everywhere it matters

**Comment by giuspen** [comment]
sorry @txe what is GMT?

**Comment by giuspen** [comment]
OK @ForeverRainbow I got the GMT, I thought it was written from @txe and didn't understand the meaning

**Comment by txe** [comment]
Updated `extension()` doesn't work right

**Comment by txe** [comment]
@giuspen , if namespace `fs` is mostly used everythere, maybe renaming CtFileSystem to `fs` is a logic step forward?

**Comment by giuspen** [comment]
@txe absolutely yes

File: future/src/ct/ct_storage_control.cc [hunk 5]
```diff
@@ -158,17 +159,17 @@ bool CtStorageControl::save(bool need_vacuum, Glib::ustring &error)
         if (need_backup)
         {
             // move is faster but the file is used by sqlite without encrypt
-            if (_file_path == _extracted_file_path && CtMiscUtil::get_doc_type(_file_path) == CtDocType::SQLite)
+            if (_file_path == _extracted_file_path && CtFileSystem::get_doc_type(_file_path) == CtDocType::SQLite)
             {
                 _storage->close_connect();    // temporary, because of sqlite keepig the file
                 if (!CtFileSystem::copy_file(_file_path, main_backup))
-                    throw std::runtime_error(str::format(_("You Have No Write Access to %s"), Glib::path_get_dirname(_file_path)));
+                    throw std::runtime_error(fmt::format(_("You Have No Write Access to {}"), _file_path.parent_path()));
```
[the remark this answers is not in the document]
Comment (ForeverRainbow) [reply future/src/ct/ct_storage_control.cc:166]: Does `fmt` not respect the translated strings? Looking at `str::format` it looks like it just replaces `%s` with `{}` and then runs it through `fmt`? Or is the `%s` needed for translation?

[the remark this answers is not in the document]
Comment (txe) [reply future/src/ct/ct_storage_control.cc:166]: %s needed for translation. Basically, any text in _() must not be changed

**Comment by giuspen** [comment]
@ForeverRainbow the _("%s") is part of the patter to be translated so you will temporarily lose the match thus that translation but you don't have to worry, the translators will fix this before the first release

**Comment by giuspen** [comment]
Sorry for the confusion I'm happy to keep the %s for now as @txe advises

File: future/tests/tests_filesystem.cpp [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+TEST(FileSystemGroup, is_directory)
+{
+    CHECK_EQUAL(true, CtFileSystem::is_directory(unitTestsDataDir));
+    CHECK_EQUAL(true, CtFileSystem::is_directory(CtFileSystem::path(unitTestsDataDir).parent_path()));
+}
+
+TEST(FileSystemGroup, get_doc_type)
+{
+    CHECK(CtDocType::SQLite == CtFileSystem::get_doc_type(unitTestsDataDir + "/test.ctb"));
```
[the remark this answers is not in the document]
Comment (ForeverRainbow) [reply future/tests/tests_filesystem.cpp:83]: Which is the 4th? I can only find `None`, `SQLite`, or `XML`?

**Comment by ForeverRainbow** [comment]
`fmt` has got a `sprintf` method btw, not sure if that helps with the translation but `str::format` could use `fmt::sprintf`?

File: future/tests/tests_filesystem.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (txe) [reply future/tests/tests_filesystem.cpp]: .ctb  .ctx  .ctd .ctz

**Comment by txe** [comment]
> fmt has got a sprintf method btw, not sure if that helps with the translation but str::format could use fmt::sprintf?

If it understands %s, then yes

**Comment by ForeverRainbow** [comment]
@txe https://fmt.dev/latest/api.html?highlight=sprintf#_CPPv2N3fmt7sprintfERK1SDpRK4Args only problem is I don't think it works with user-defined types (i.e `Glib::ustring`)

**Comment by txe** [comment]
I think we can just fix all translated string by replacing %s by {} in code and translation files. So for now we can left str::format as it is

File: future/tests/tests_filesystem.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (ForeverRainbow) [reply future/tests/tests_filesystem.cpp]: Sorry your comments are lagging for me, only just got that

**Comment by txe** [comment]
The tests for extension looks really great :)
So far code looks OK, but it is hard to say by just reviewing diffs.
I'm going to take your branch and look through by IDE. Maybe run code on win. It takes this day.

**Comment by txe** [comment]
Hi, there are some issues,
Maybe the process will go faster if I add myself changes in your branch?

**Comment by ForeverRainbow** [comment]
@txe Sure that would be great!

**Comment by txe** [comment]
remove exceptions because
- they don't help much in this case and code actually doesn't catch them
- it's confusing, a couple functions throw exceptions but others not

remove fs::append, operator path::/= 
- they don't behave like std::filesystem analogs
- std::filesystem analogs are really confusing, it's better to avoid using them at all

**Pull request merged**